src/Microsoft.VisualStudio.Services.Agent/ProcessWrapper.cs

```diff
@@ -0,0 +1,295 @@
+        {
+            m_asyncStreamReaderCount++;
+            Task.Run(() => ReadStream(reader, dataHandler)).
+                ContinueWith((t)=>ReadStreamComplete(t));
+        }
+
+        private void ReadStreamComplete(IAsyncResult result)
+        {
+            // Don't access m_asyncStreamReaderCount or run the exited handler until starting is complete
+            lock (Lock)
+            {
+                if (Interlocked.Decrement(ref m_asyncStreamReaderCount) == 0 && m_waitingOnStreams)
+                {
+                    CallProcessExitCallback(EventArgs.Empty);
+                }
+            }
+        }
+
+        private delegate void ReadStreamHandler(StreamReader reader, EventHandler<ProcessDataReceivedEventArgs> dataHandler);
+        private ReadStreamHandler ReadStream { get; set; }
+
+        private void ReadStreamCore(StreamReader reader, EventHandler<ProcessDataReceivedEventArgs> dataHandler)
+        {
+            while (!reader.EndOfStream)
+            {
+                String line = reader.ReadLine();
+
+                if (dataHandler != null)
+                {
+                    dataHandler(Process, new ProcessDataReceivedEventArgs(line));
+                }
+            }
+        }
+
+        private void ProcessExitedHandler(object sender, EventArgs e)
+        {
+            if ((StartInfo.RedirectStandardError || StartInfo.RedirectStandardOutput) && m_asyncStreamReaderCount != 0)
+            {
+                m_waitingOnStreams = true;
+
+                Task.Run(async delegate
+                {
+                    // Wait 5 seconds and then kill everythin in the Job just in case
+                    await Task.Delay(TimeSpan.FromSeconds(5));
+                    CallProcessExitCallback(e);
+                });
+            }
+            else
+            {
+                CallProcessExitCallback(e);
+            }
+        }
+
+        private void CallProcessExitCallback(EventArgs e)
+        {
+            lock (Lock)
+            {
+                if (m_exited)
+                {
+                    return;
+                }
+
+                m_exited = true;
+                ProcessExitedEvent.TrySetResult(true);
+            }
+
+            var exitHandler = this.Exited;
+            if (exitHandler != null)
+            {
+                exitHandler(this, e);
+            }
+        }
+
+        private void ErrorDataHandler(object sender, ProcessDataReceivedEventArgs e)
+        {
+            if (ErrorDataReceived != null)
+            {
+                ErrorDataReceived(this, e);
+            }
+        }
+
+        private void OutputDataHandler(object sender, ProcessDataReceivedEventArgs e)
+        {
+            if (OutputDataReceived != null)
+            {
+                OutputDataReceived(this, e);
+            }
+        }
+    }
+
+    public class ProcessDataReceivedEventArgs : EventArgs
```

Comment:
Can this be deleted? Isn't this interface the same as `DataReceivedEventArgs`, with the exception of the constructor?